lib/cc/analyzer/source_extractor.rb

```diff
@@ -0,0 +1,66 @@
+module CC
+  module Analyzer
+    class SourceExtractor
+      def initialize(source)
+        @source = source
+      end
+
+      def extract(location)
+        case
+        when (lines = location["lines"])
+          extract_from_lines(lines)
+        when (positions = location["positions"])
+          extract_from_positions(positions)
+        end
+      end
+
+      private
+
+      attr_reader :location, :source
+
+      def extract_from_lines(lines)
+        range = (lines["begin"] - 1..lines["end"] - 1)
+
+        source.each_line.with_object("").with_index do |(line, memo), index|
+          memo << line if range.include?(index)
+        end
+      end
+
+      def extract_from_positions(positions)
+        positions = convert_to_offsets(positions)
+
+        length = positions["end"]["offset"] - positions["begin"]["offset"]
```

Comment:
The nested hash access seems ripe for `nil`-errors. Are you confident these keys will be here? Would `Hash#fetch` be too syntactically noisy?